doc/source/jupyterhub/customizing/user-management.md

```diff
@@ -17,7 +17,7 @@ will be lost. This is known as _culling_.
 
 ```{note}
 In JupyterHub, "inactivity" is defined as no response from the user's
-browser. JupyterHub constantly pings the user's JupyterHub browser session via the JupyterHub API
+browser. The `jupyterhub-idle-culler` running on the `hub` pod constantly pings the user's JupyterHub browser session via the JupyterHub API
```

Comment:
I think this is clear... if wordy.

Reply:
> pings the user's JupyterHub browser session via the JupyterHub API

Hmm, what is the JupyterHub API used specifically? I assume it isn't directly speaking with the user's browsers, but the user server.

### This is my understanding overviewed

- Idle culler knows JupyterHub
- JupyterHub knows the user server
- The user server knows the kernels and can know about requests from the browser


I lack understanding of jupyterhub-idle-culler's function, does it for example use the JupyterHub API to get information about running servers and then pings them by itself, or does it ask JupyterHub to poll the user servers for some updated information or similar? I think it asks for information about the user server, and then polls the user server but I'm not sure.

Reply:
Hm yes let me try to get this right. I see that `jupyterhub-idle-culler` (running on the Hub pod) is polling the Hub API for `users`, which is a GET endpoint that returns the user info but also the list of that user's servers along with a timestamp when they were launched, a `last_activity` timestamp, and some other status information. Here's an example:

```
{'kind': 'user', 'name': 'cdibble', 'admin': True, 'groups': [], 'server': '/user/cdibble/', 'pending': None, 'created': '2021-06-18T19:41:52.783461Z', 'last_activity': '2021-06-21T22:11:53.697681Z', 'servers': {'': {'name': '', 'last_activity': '2021-06-21T22:10:55.225000Z', 'started': '2021-06-21T19:42:50.076031Z', 'pending': None, 'ready': True, 'state': {'pod_name': 'jupyter-cdibble'}, 'url': '/user/cdibble/', 'user_options': {'profile': 'fry-2-vcpu-7gb-ram'}, 'progress_url': '/hub/api/users/cdibble/server/progress'}}}
```
That information is available because of regular POSTs to the Hub API, but those actually seem to originate from the Hub pod as well. So I suppose that `jupyterhub-idle-culler` is relying on JupyterHub to poll the user servers.

Incidentally, it does not currently ask for IPs (which you can get with the `/proxy` endpoint on the JupyterHub API). Using that endpoint to query user pods' IPs is part of how I am hoping change the culling behavior a bit.

So the above sentence might better read:

```
The `jupyterhub-idle-culler` running on the `hub` pod routinely pings JupyterHub API for information about user server activity, which is made available because JupyterHub is constantly polling kernels for activity and status information.
```

Reply:
What do you think @consideRatio?

Reply:
@cdibble excellent investigation thank you for clearing this up!

I guess JupyterHub polls the user server rather than the kernels directly, and that the user server provides a summary based on its knowledge. How the user server acquired info about kernel activity is out of scope to discuss I think.